src/app/ui/sequence.tsx

```diff
@@ -0,0 +1,860 @@
+        if (opKey(l) !== operatorKey) continue;
+
+        units.push(unit);
+    }
+
+    if (units.length > 0) {
+        const data = { structure, units };
+        const unit = units[0];
+
+        let sw: SequenceWrapper<any>;
+        if (unit.polymerElements.length) {
+            const l = StructureElement.Location.create(structure, unit, unit.elements[0]);
+            const entitySeq = unit.model.sequence.byEntityKey[SP.entity.key(l)];
+            // check if entity sequence is available
+            if (entitySeq && entitySeq.sequence.length <= MaxDisplaySequenceLength) {
+                sw = new PolymerSequenceWrapper(data);
+            } else {
+                const polymerElementCount = units.reduce((a, v) => a + v.polymerElements.length, 0);
+                if (Unit.isAtomic(unit) || polymerElementCount > MaxDisplaySequenceLength) {
+                    sw = new ChainSequenceWrapper(data);
+                } else {
+                    sw = new ElementSequenceWrapper(data);
+                }
+            }
+        } else if (Unit.isAtomic(unit)) {
+            const residueCount = units.reduce((a, v) => a + (v as Unit.Atomic).residueCount, 0);
+            if (residueCount > MaxDisplaySequenceLength) {
+                sw = new ChainSequenceWrapper(data);
+            } else {
+                sw = new HeteroSequenceWrapper(data);
+            }
+        } else {
+            console.warn('should not happen, expecting coarse units to be polymeric');
+            sw = new ChainSequenceWrapper(data);
+        }
+
+        sw.markResidue(structureSelection.getLoci(structure), MarkerAction.Select);
+        return sw;
+    } else {
+        return 'No sequence available';
+    }
+}
+
+export function getLigand(options: { chainId: string; ligandId: string }, structure: Structure) {
+    const ligands = structure.units.flatMap((unit: Unit) =>
+        Array.from(unit.elements).flatMap(element => {
+            const location = {
+                kind: "element-location" as const,
+                structure,
+                unit,
+                element,
+            };
+
+            const compIds = StructureProperties.residue.microheterogeneityCompIds(location);
+            const type = StructureProperties.residue.group_PDB(location);
+            const authSeqId = StructureProperties.residue.auth_seq_id(location);
+            const chainId = StructureProperties.chain.auth_asym_id(location);
+            const structAsymId = StructureProperties.chain.label_asym_id(location);
+
+            // Debug ligands without previous filtering
+            // const ligand = [compIds[0], authSeqId].join("-");
+
+            return { type, compIds, authSeqId, chainId, structAsymId, id: [compIds[0], authSeqId].join("-") };
+        })
+    );
+
+    const groupedLigandsById = _.groupBy(ligands, 'id');
+    return groupedLigandsById[options.ligandId].filter(ligand => ligand.chainId === options.chainId)[0];
```

Comment:
`groupedLigandsById[options.ligandId]` could be undefined. Ts is not showing the problem because flag noUncheckedIndexedAccess in tsconfig is disabled.

Reply:
Oh! nice catch! Thank you!